compiler/src/dotty/tools/dotc/transform/init/Checking.scala

```diff
@@ -254,34 +254,98 @@ object Checking {
+    }
+
+  /// Check if we can just directly promote a potential.
+  /// A potential can be (currently) directly promoted if and only if:
+  /// - `pot == this` and all fields of this are initialized, or
+  /// - `pot == Warm(C, outer)` where `outer` can be directly promoted.
+  private def canDirectlyPromote(pot: Potential)(using state: State): Boolean = trace("checking direct promotion of " + pot.show, init) {
+    if (state.safePromoted.contains(pot)) true
+    else pot match {
+      case pot: ThisRef =>
+        // If we have all fields initialized, then we can promote This to hot.
+        val classRef = state.thisClass.info.asInstanceOf[ClassInfo].appliedRef
+        classRef.fields.forall { denot =>
+          val sym = denot.symbol
+          sym.isOneOf(Flags.Lazy | Flags.Deferred) || state.fieldsInited.contains(sym)
+        }
+      case Warm(cls, outer) =>
+        canDirectlyPromote(outer)
+      case _ =>
+        val summary = expand(pot)
+        if (!summary.effs.isEmpty)
+          false // max depth of expansion reached
+        else summary.pots.forall(canDirectlyPromote)
+    }
+  }
+
+  /// Check the Promotion of a Warm object, according to "Rule 2":
+  //
+  // Rule 2: Promote(pot)
+  //
+  // for all concrete methods `m` of D
+  //    pot.m!, Promote(pot.m)
+  //
+  // for all concrete fields `f` of D
+  //    Promote(pot.f)
+  //
+  // for all inner classes `F` of D
+  //    Warm[F, pot].init!, Promote(Warm[F, pot])
+  private def checkPromoteWarm(warm: Warm, eff: Effect)(using state: State): Errors =
+    val Warm(cls, outer) = warm
+    val source = eff.source
+    // Errors.empty
+    val classRef = cls.info.asInstanceOf[ClassInfo].appliedRef
+    // All members of class must be promotable.
+    val buffer = new mutable.ArrayBuffer[Effect]
+    val excludedFlags = Flags.Deferred | Flags.Private | Flags.Protected
+
+    classRef.fields.foreach { denot =>
+      val f = denot.symbol
+      if !f.isOneOf(excludedFlags) && f.hasSource then
+        buffer += Promote(FieldReturn(warm, f)(source))(source)
+        buffer += FieldAccess(warm, f)(source)
+    }
 
+    classRef.membersBasedOnFlags(Flags.Method, Flags.Deferred).foreach { denot =>
+      val m = denot.symbol
+      if !m.isConstructor && m.hasSource && !theEnv.canIgnoreMethod(m) then
+        buffer += MethodCall(warm, m)(source)
+        buffer += Promote(MethodReturn(warm, m)(source))(source)
     }
 
+    classRef.memberClasses.foreach { denot =>
+      val cls = denot.symbol.asClass
+      if cls.hasSource then
+        val potInner = Potentials.asSeenFrom(Warm(cls, ThisRef()(source))(source), warm)
+        buffer += MethodCall(potInner, cls.primaryConstructor)(source)
+        buffer += Promote(potInner)(source)
+    }
+
+    val errs = buffer.toList.flatMap(eff => check(eff))
+    if errs.isEmpty then {
+      Errors.empty
+    } else {
+      UnsafePromotion(warm, eff.source, state.path, errs.toList).toErrors
+    }
```

Comment:
Minor: the curly braces are not necessary here (and Scala 3 makes them always optional).